server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentKiller.java

```diff
@@ -0,0 +1,136 @@
+import org.joda.time.Interval;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ */
+public class DruidCoordinatorSegmentKiller implements DruidCoordinatorHelper
+{
+  private final static Logger log = new Logger(DruidCoordinatorSegmentKiller.class);
+
+  private final long period;
+  private final long retainDuration;
+  private final int maxSegmentsToKill;
+  private long lastKillTime = 0;
+
+
+  private final MetadataSegmentManager segmentManager;
+  private final IndexingServiceClient indexingServiceClient;
+
+  public DruidCoordinatorSegmentKiller(
+      MetadataSegmentManager segmentManager,
+      IndexingServiceClient indexingServiceClient,
+      Duration retainDuration,
+      Duration period,
+      int maxSegmentsToKill
+  )
+  {
+    this.period = period.getMillis();
+    Preconditions.checkArgument(this.period > 0, "coordinator kill period must be > 0");
+
+    this.retainDuration = retainDuration.getMillis();
+    Preconditions.checkArgument(this.retainDuration >= 0, "coordinator kill retainDuration must be >= 0");
+
+    this.maxSegmentsToKill = maxSegmentsToKill;
+    Preconditions.checkArgument(this.maxSegmentsToKill > 0, "coordinator kill maxSegments must be > 0");
+
+    log.info(
+        "Kill Task scheduling enabled with period [%s], retainDuration [%s], maxSegmentsToKill [%s]",
+        this.period,
+        this.retainDuration,
+        this.maxSegmentsToKill
+    );
+
+    this.segmentManager = segmentManager;
+    this.indexingServiceClient = indexingServiceClient;
+  }
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+  {
+    Set<String> whitelist = params.getCoordinatorDynamicConfig().getKillDataSourceWhitelist();
+
+    if (whitelist != null && whitelist.size() > 0 && (lastKillTime + period) < System.currentTimeMillis()) {
+      lastKillTime = System.currentTimeMillis();
+
+      for (String dataSource : whitelist) {
+        final Interval intervalToKill = findIntervalForKillTask(dataSource, maxSegmentsToKill);
+        if (intervalToKill != null) {
+          try {
+            indexingServiceClient.killSegments(dataSource, intervalToKill);
+          }
+          catch (Exception ex) {
+            log.error(ex, "Failed to submit kill task for dataSource [%s]", dataSource);
+            if (Thread.currentThread().isInterrupted()) {
+              log.warn("skipping kill task scheduling because thread is interrupted.");
+              break;
+            }
+          }
+        }
+      }
+    }
+    return params;
+  }
+
+  @VisibleForTesting
+  Interval findIntervalForKillTask(String dataSource, int limit)
+  {
+    List<Interval> unusedSegmentIntervals = segmentManager.getUnusedSegmentIntervals(
+        dataSource,
+        new Interval(
+            0,
+            System.currentTimeMillis()
+            - retainDuration
+        ),
+        limit
+    );
+
+    if (unusedSegmentIntervals != null && unusedSegmentIntervals.size() > 0) {
```

Comment:
could this be replaced with return JodaUtils.umbrellaInterval(usedSegmentIntervals) ?